posix/unix.c

```diff
@@ -547,59 +640,144 @@ ssize_t unix_sendto(unsigned socket, const void *message, size_t length, int fla
 
 	do {
 		if (s->type == SOCK_DGRAM) {
-			conn = s;
-		}
-		else if (s->connect == NULL && dest_addr != NULL) {
-			if ((err = unix_lookupSocket(dest_addr->sa_data)) < 0)
-				break;
+			if (dest_addr && dest_len != 0) {
+				if ((err = unix_lookupSocket(dest_addr->sa_data)) < 0)
+					break;
 
-			if ((conn = unixsock_get(err)) == NULL) {
-				err = -ENOTSOCK;
+				if ((conn = unixsock_get(err)) == NULL) {
+					err = -ENOTSOCK;
+					break;
+				}
+			}
+			else if ((conn = s->connect) == NULL) {
+				err = -ENOTCONN;
 				break;
 			}
 		}
-		else if ((conn = s->connect) == NULL) {
-			err = -ENOTCONN;
-			break;
+		else {
+			if (dest_addr || dest_len != 0) {
+				err = -EISCONN;
+				break;
+			}
+			else if ((conn = s->connect) == NULL) {
+				err = -ENOTCONN;
+				break;
+			}
 		}
 
 		err = 0;
-		for (;;) {
-			proc_lockSet(&conn->lock);
-			if (s->type == SOCK_STREAM) {
-				err = _cbuffer_write(&conn->buffer, message, length);
-			}
-			else if (_cbuffer_free(&conn->buffer) >= length + sizeof(length)) {
-				_cbuffer_write(&conn->buffer, &length, sizeof(length));
-				_cbuffer_write(&conn->buffer, message, err = length);
-				proc_threadWakeup(&conn->queue);
-			}
-			proc_lockClear(&conn->lock);
 
-			if (err > 0) {
+		if (len > 0) {
+			for (;;) {
+				proc_lockSet(&conn->lock);
+				if (s->type == SOCK_STREAM) {
+					err = _cbuffer_write(&conn->buffer, buf, len);
+				}
+				else if (_cbuffer_free(&conn->buffer) >= len + sizeof(len)) { /* SOCK_DGRAM or SOCK_SEQPACKET */
+					_cbuffer_write(&conn->buffer, &len, sizeof(len));
+					_cbuffer_write(&conn->buffer, buf, err = len);
+					proc_threadWakeup(&conn->queue);
```

Comment:
Missing spinlock

Reply:
This wakeup seems to be redundant so I've removed it.